orderbook/src/solvable_orders.rs

```diff
@@ -0,0 +1,379 @@
+use crate::{
+    account_balances::{BalanceFetching, Query},
+    database::orders::OrderStoring,
+    orderbook::filter_unsupported_tokens,
+};
+use anyhow::Result;
+use model::order::Order;
+use primitive_types::U256;
+use shared::{
+    bad_token::BadTokenDetecting, current_block::CurrentBlockStream, time::now_in_epoch_seconds,
+};
+use std::{
+    collections::{HashMap, HashSet},
+    iter::FromIterator,
+    sync::{Arc, Mutex, Weak},
+    time::{Duration, Instant},
+};
+use tokio::sync::Notify;
+
+/// Keeps track and updates the set of currently solvable orders.
+/// For this we also need to keep track of user sell token balances for open orders so this is
+/// retrievable as well.
+/// The cache is updated in the background whenever a new block appears or when the cache is
+/// explicitly notified that it should update for example because a new order got added to the order
+/// book.
+pub struct SolvableOrdersCache {
+    min_order_validity_period: Duration,
+    database: Arc<dyn OrderStoring>,
+    balance_fetcher: Arc<dyn BalanceFetching>,
+    bad_token_detector: Arc<dyn BadTokenDetecting>,
+    notify: Notify,
+    cache: Mutex<Inner>,
+}
+
+type Balances = HashMap<Query, U256>;
+
+struct Inner {
+    orders: Vec<Order>,
+    balances: Balances,
+    block: u64,
+    update_time: Instant,
```

Comment:
Would it be more appropriate to use `update_block` instead of a time? I ask because I imagine we would be comparing a block number with the last update to determine if an update it necessary.

Reply:
It uses to block whether it should requery  balances. The update time is used in the orderbook to figure out when we should enter an error state if the solvable orders haven't been updated in X seconds real time.